### PR TITLE
(enhancement)engine_status_patch:  enhance engine status patch logic based on updated schema

### PR DESCRIPTION
**Pull request by ksatchit**
Signed-off-by: ksatchit <karthik.s@mayadata.io>

**What this PR does / why we need it**:

- Patches the engine status based on updated schema proposed in https://github.com/litmuschaos/chaos-operator/pull/251
- This PR facilitates other litmus components (portal) and external tools to parse readily usable info from the ChaosEngine status (this includes experiment names, chaos-runenr pod name, experiment pod name) 
- Considering we use & project the ChaosEngine as the single source of truth, the status should be more informative rather than provide hints that forces developers to _derive_ data. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes # https://github.com/litmuschaos/chaos-operator/issues/250

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

**Comment by ksatchit** [comment]
## TestCases

### ChaosEngine Status Over Standard/Regular Lifecycle of Experiment Execution

- Immediately after creation of chaos-runner & before launch of experiment pods 

```yaml
Status:
  Engine Status:  initialized
  Experiments:
    Experiment Pod:    Yet to be launched
    Last Update Time:  2020-07-31T13:35:43Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            Waiting for Job Creation
    Verdict:           N/A
```
- During experiment execution

```yaml
Status:
  Engine Status:  initialized
  Experiments:
    Experiment Pod:    pod-delete-2axi7k-sdnkf
    Last Update Time:  2020-07-31T13:35:55Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            Running
    Verdict:           Awaited
```

- After experiment execution has succeeded, before decision on retention/cleanup of chaos resources 

```yaml
Status:
  Engine Status:  initialized
  Experiments:
    Experiment Pod:    pod-delete-2axi7k-sdnkf
    Last Update Time:  2020-07-31T13:36:35Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            Execution Successful
    Verdict:           Pass
```

- After implementing the JobCleanUpPolicy

```yaml
Status:
  Engine Status:  completed
  Experiments:
    Experiment Pod:    pod-delete-xauv6k-lrn2c
    Last Update Time:  2020-07-31T13:40:11Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            Execution Successful
    Verdict:           Pass
```

### Failure of experiment via genuine (via exit conditions) causes 

```yaml
Status:
  Engine Status:  completed
  Experiments:
    Experiment Pod:    pod-delete-o3vd7r-mpmtm
    Last Update Time:  2020-07-31T13:49:04Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            Execution Successful
    Verdict:           Fail
```

### Failure to launch experiment pods due to lack of experiment resources

```yaml
Status:
  Engine Status:  completed
  Experiments:
    Experiment Pod:    N/A
    Last Update Time:  2020-07-31T13:59:26Z
    Name:              pod-delete
    Runner:            nginx-chaos-runner
    Status:            ChaosExperiment Not Found
    Verdict:           Failed
```

File: pkg/utils/watchChaosContainer.go [hunk 1]
```diff
@@ -2,32 +2,57 @@ package utils
 
 import (
 	"time"
-
```
Comment (chandankumar4) [inline pkg/utils/watchChaosContainer.go:5]: This line gap is required!

File: pkg/utils/watchChaosContainer.go [hunk 2]
```diff
@@ -2,32 +2,57 @@ package utils
...
 )
 
+
+var (
+	//chaosPod holds the experiment pod spec
+	chaosPod = &corev1.Pod{}
+)
+
+
+// GetChaosPodName gets the name of the chaos experiment pod launched by the runner
```
Comment (chandankumar4) [inline pkg/utils/watchChaosContainer.go:17]: Comment is not fit a/c to method definition

File: pkg/utils/watchChaosContainer.go [hunk 3]
```diff
@@ -2,32 +2,57 @@ package utils
...
-
 	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+
+var (
+	//chaosPod holds the experiment pod spec
+	chaosPod = &corev1.Pod{}
```
Comment (chandankumar4) [inline pkg/utils/watchChaosContainer.go:13]: Are we using this global definition somewhere? better approach is to define this inside the method itself

File: pkg/utils/watchChaosContainer.go [hunk 4]
```diff
@@ -2,32 +2,57 @@ package utils
...
+
+// GetChaosPodName gets the name of the chaos experiment pod launched by the runner
+func GetChaosPod(expDetails *ExperimentDetails, clients ClientSets) (*corev1.Pod, error){
+	chaosPodList, err := clients.KubeClient.CoreV1().Pods(expDetails.Namespace).List(metav1.ListOptions{LabelSelector: "job-name=" + expDetails.JobName})
+	if err != nil || len(chaosPodList.Items) == 0 {
+    	return nil, errors.Wrapf(err, "Unable to get the chaos pod, due to error: %v", err)
+	} else if len(chaosPodList.Items) > 1 {
+		// Cases where experiment pod is rescheduled by the job controller due to 
+		// issues while the older pod is still not cleaned-up
+		return nil, errors.New("Multiple pods exist with same jobname label")
```
Comment (chandankumar4) [inline pkg/utils/watchChaosContainer.go:25]: Can we log the name of the experiment also or some more detail for debug it easily?

File: pkg/utils/watchChaosContainer.go [hunk 5]
```diff
@@ -49,7 +75,12 @@ func (engineDetails EngineDetails) WatchChaosContainerForCompletion(experiment *
 		}
 
 		var expStatus ExperimentStatus
-		expStatus.AwaitedExperimentStatus(experiment)
+        chaosPod, err := GetChaosPod(experiment, clients)
```
Comment (chandankumar4) [inline pkg/utils/watchChaosContainer.go:78]: Indentation is not correct here!

File: pkg/utils/watchJob.go [hunk 6]
```diff
@@ -75,7 +75,11 @@ func (engineDetails EngineDetails) UpdateEngineWithResult(experiment *Experiment
 	}
 
 	var currExpStatus ExperimentStatus
-	currExpStatus.CompletedExperimentStatus(chaosResult, experiment)
+    chaosPod, err := GetChaosPod(experiment, clients)
```
Comment (chandankumar4) [inline pkg/utils/watchJob.go:78]: Indentation is not correct

File: pkg/utils/watchChaosContainer.go (hunk 5 above)
Comment (ksatchit) [reply pkg/utils/watchChaosContainer.go]: Done!

File: pkg/utils/watchJob.go (hunk 6 above)
Comment (ksatchit) [reply pkg/utils/watchJob.go]: Done!

File: pkg/utils/watchChaosContainer.go (hunk 2 above)
Comment (ksatchit) [reply pkg/utils/watchChaosContainer.go]: Updated!

File: pkg/utils/status.go [hunk 7]
```diff
@@ -6,35 +6,43 @@ import (
...
 
 // CompletedExperimentStatus fills up ExperimentStatus Structure with values chaosResult
-func (expStatus *ExperimentStatus) CompletedExperimentStatus(chaosResult *v1alpha1.ChaosResult, experimentDetails *ExperimentDetails) {
+func (expStatus *ExperimentStatus) CompletedExperimentStatus(chaosResult *v1alpha1.ChaosResult, engineName, experimentPodName string) {
 	//var currExpStatus v1alpha1.ExperimentStatuses
-	expStatus.Name = experimentDetails.JobName
+	expStatus.Name = chaosResult.Spec.ExperimentName
+	expStatus.Runner = engineName + "-runner"
+	expStatus.ExpPod = experimentPodName
 	expStatus.Status = "Execution Successful"
```
Comment (ispeakc0de) [inline pkg/utils/status.go:34]: experiment status in chaosengine `Completed` vs `Execution Successful`?

Comment (ksatchit) [reply pkg/utils/status.go]: Makes sense! Thanks

File: pkg/utils/watchChaosContainer.go (hunk 3 above)
Comment (ksatchit) [reply pkg/utils/watchChaosContainer.go]: Done!

File: pkg/utils/watchChaosContainer.go (hunk 4 above)
Comment (ksatchit) [reply pkg/utils/watchChaosContainer.go:19]: Considering we are doing a 1-1 map b/w engine and experiment, this should be easily debuggable. Even if there are more than 1 experiments listed in the engine, their execution is sequential with events showing exp that has failed.

Comment (chandankumar4) [reply pkg/utils/watchChaosContainer.go:19]: ok

File: pkg/utils/watchJob.go [hunk 8]
```diff
@@ -75,7 +75,11 @@ func (engineDetails EngineDetails) UpdateEngineWithResult(experiment *Experiment
 	}
 
 	var currExpStatus ExperimentStatus
-	currExpStatus.CompletedExperimentStatus(chaosResult, experiment)
+	chaosPod, err := GetChaosPod(experiment, clients)
+    if err != nil {
```
Comment (chandankumar4) [inline pkg/utils/watchJob.go:79]: Indentation

**Pull request merged**